src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,449 @@
+{ default := ⟨⊥, ideal.is_homogeneous.bot _⟩}
+
+instance homogeneous_ideal.has_top :
+  has_top (homogeneous_ideal 𝒜) :=
+⟨⟨⊤, λ _ _ _, by simp only [submodule.mem_top]⟩⟩
+
+@[simp] lemma homogeneous_ideal.eq_top_iff (I : homogeneous_ideal 𝒜) : I = ⊤ ↔ I.1 = ⊤ :=
+subtype.ext_iff
+
+instance homogeneous_ideal.order : partial_order (homogeneous_ideal 𝒜) :=
+partial_order.lift _ subtype.coe_injective
+
+instance homogeneous_ideal.has_mem : has_mem A (homogeneous_ideal 𝒜) :=
+{ mem := λ r I, r ∈ I.1 }
+
+variables {𝒜}
+
+lemma ideal.is_homogeneous.inf {I J : ideal A}
+  (HI : I.is_homogeneous 𝒜) (HJ : J.is_homogeneous 𝒜) : (I ⊓ J).is_homogeneous 𝒜 :=
+λ i r hr, ⟨HI _ hr.1, HJ _ hr.2⟩
+
+lemma ideal.is_homogeneous.Inf {ℐ : set (ideal A)} (h : ∀ I ∈ ℐ, ideal.is_homogeneous 𝒜 I) :
+  (Inf ℐ).is_homogeneous 𝒜 :=
+begin
+  intros i x Hx, simp only [ideal.mem_Inf] at Hx ⊢,
+  intros J HJ,
+  exact h _ HJ _ (Hx HJ),
+end
+
+lemma ideal.is_homogeneous.mul {I J : ideal A}
+  (HI : I.is_homogeneous 𝒜) (HJ : J.is_homogeneous 𝒜) : (I * J).is_homogeneous 𝒜 :=
+begin
+  rw ideal.is_homogeneous.iff_exists at HI HJ ⊢,
+  obtain ⟨⟨s₁, rfl⟩, ⟨s₂, rfl⟩⟩ := ⟨HI, HJ⟩,
+  rw ideal.span_mul_span',
+  refine ⟨s₁ * s₂, congr_arg _ _⟩,
+  exact (set.image_mul (submonoid.subtype _).to_mul_hom).symm,
+end
+
+lemma ideal.is_homogeneous.sup {I J : ideal A}
+  (HI : I.is_homogeneous 𝒜) (HJ : J.is_homogeneous 𝒜) : (I ⊔ J).is_homogeneous 𝒜 :=
+begin
+  rw ideal.is_homogeneous.iff_exists at HI HJ ⊢,
+  obtain ⟨⟨s₁, rfl⟩, ⟨s₂, rfl⟩⟩ := ⟨HI, HJ⟩,
+  refine ⟨s₁ ∪ s₂, _⟩,
+  rw [set.image_union],
+  exact (submodule.span_union _ _).symm,
+end
+
+lemma ideal.is_homogeneous.Sup
+  {ℐ : set (ideal A)} (Hℐ : ∀ (I ∈ ℐ), ideal.is_homogeneous 𝒜 I) :
+  (Sup ℐ).is_homogeneous 𝒜 :=
+begin
+  simp_rw ideal.is_homogeneous.iff_exists at Hℐ ⊢,
+  choose 𝓈 h𝓈 using Hℐ,
+  refine ⟨⋃ I hI, 𝓈 I hI, _⟩,
+  simp_rw [set.image_Union, ideal.span_Union, Sup_eq_supr],
+  conv in (ideal.span _) { rw ←h𝓈 i x },
+end
+
+variables (𝒜)
+
+instance : has_inf (homogeneous_ideal 𝒜) :=
+{ inf := λ I J, ⟨I ⊓ J, I.prop.inf J.prop⟩ }
+
+instance : has_Inf (homogeneous_ideal 𝒜) :=
+{ Inf := λ ℐ, ⟨Inf (coe '' ℐ), ideal.is_homogeneous.Inf $ λ _ ⟨I, _, hI⟩, hI ▸ I.prop⟩ }
+
+instance : has_sup (homogeneous_ideal 𝒜) :=
+{ sup := λ I J, ⟨I ⊔ J, I.prop.sup J.prop⟩ }
+
+instance : has_Sup (homogeneous_ideal 𝒜) :=
+{ Sup := λ ℐ, ⟨Sup (coe '' ℐ), ideal.is_homogeneous.Sup $ λ _ ⟨I, _, hI⟩, hI ▸ I.prop⟩ }
+
+lemma homogeneous_ideal.Sup_eq (ℐ : set (homogeneous_ideal 𝒜)) :
+  ((Sup ℐ : homogeneous_ideal 𝒜) : ideal A) = Sup ((coe : homogeneous_ideal 𝒜 → ideal A) '' ℐ) :=
+rfl
+
+lemma homogeneous_ideal.supr_eq (s : ι → homogeneous_ideal 𝒜) :
+  (((⨆ (i : ι), s i) : homogeneous_ideal 𝒜) : ideal A) = ⨆ i, (s i : ideal A) :=
+begin
+  unfold supr,
+  rw homogeneous_ideal.Sup_eq,
+  congr,
+  ext I,
+  split,
+  { rintro ⟨_, ⟨i, -, rfl⟩, rfl⟩,
+    exact ⟨i, rfl⟩, },
+  { rintro ⟨i, -, rfl⟩,
+    exact ⟨s i, ⟨⟨i, rfl⟩, rfl⟩⟩, }
+end
```

Comment:
There ought to be a really short proof for this - maybe have a look at some other `coe_supr` lemmas to see if there's a trick - or wait for another reviewer to golf it!

Reply:
I made it slightly shorter
```lean
lemma homogeneous_ideal.coe_supr (s : ι → homogeneous_ideal 𝒜) :
  ↑(⨆ i, s i) = ⨆ i, (s i : ideal A) :=
let h : ∀ (I : ideal A), I ∈ (coe : homogeneous_ideal 𝒜 → ideal A) '' range (λ (i : ι), s i) ↔
  I ∈ range ((coe : homogeneous_ideal 𝒜 → ideal A) ∘ s) := λ I,
    ⟨ by { rintro ⟨_, ⟨i, -, rfl⟩, rfl⟩, exact ⟨i, rfl⟩ },
      by { rintro ⟨i, -, rfl⟩, exact ⟨s i, ⟨⟨i, rfl⟩, rfl⟩⟩ }⟩ in
eq.trans (homogeneous_ideal.coe_Sup _ _) $ congr_arg _ (ext h)
```

Reply:
I ended up pushing:
```lean
@[simp] lemma homogeneous_ideal.coe_supr {ι' : Sort*} (s : ι' → homogeneous_ideal 𝒜) :
  ↑(⨆ i, s i) = ⨆ i, (s i : ideal A) :=
by rw [supr, supr, homogeneous_ideal.coe_Sup, ←set.range_comp]
```
which is the type of thing I was hoping for. If you're resorting  to `rintro` to clean up a set mess, you've probably missed a lemma (in this case, `set.range_comp`).